### PR TITLE
[SPARK-34061][SQL] Support push down DISTINCT in Set operations

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR makes it support push down DISTINCT in Set operations(`INTERSECT`/`EXCEPT`) if push down has benefit.

### Why are the changes needed?

Improve `INTERSECT`/`EXCEPT` performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and TPC-DS benchmark test with (`spark.sql.adaptive.enabled=false`).

SQL | Before this PR(Seconds) | After this PR(Seconds)
-- | -- | --
q14a | 396  | 78
q14b | 378  | 78
q38 | 29 | 28
q87 | 35 | 28

### CODE REVIEW

**Review by blaquezzz — APPROVED** [review]

**Comment by tanelk** [comment]
Does it work for `ReplaceExceptWithAntiJoin` also?

**Comment by wangyum** [comment]
> Does it work for `ReplaceExceptWithAntiJoin` also?

Yes, it should work for `ReplaceExceptWithAntiJoin`,  but it doesn't improve TPC-DS queries. So I didn't add it to the pr.

**Comment by dongjoon-hyun** [comment]
Sorry for being late, @wangyum . Could you rebase this PR to the master?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1663,8 +1663,12 @@ object ReplaceDeduplicateWithAggregate extends Rule[LogicalPlan] {
 /**
  * Replaces logical [[Intersect]] operator with a left-semi [[Join]] operator.
  * {{{
- *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2
- *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
+ *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2 ==>
+ *   SELECT a1, a2 FROM
+ *     (SELECT DISTINCT a1, a2 FROM Tab1)
+ *   LEFT SEMI JOIN
+ *     (SELECT DISTINCT b1, b2 FROM Tab2)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1670]: Do you have any reference from the other DBMSs?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1663,8 +1663,12 @@ object ReplaceDeduplicateWithAggregate extends Rule[LogicalPlan] {
 /**
  * Replaces logical [[Intersect]] operator with a left-semi [[Join]] operator.
  * {{{
- *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2
- *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1<=>b1 AND a2<=>b2
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1667]: This seems like we push down the aggregate operator through the join operator. It's usually beneficial, but can cause perf regression if the join children do not have many duplications and the aggregate operator can't reduce the data volume.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1667]: Could we add a config?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1667]: Another similar case.
Origin SQL:
```sql
SELECT DISTINCT t.at_prod_ref_id, d.week_beg_dt
FROM   table1 t
       JOIN table2 d
         ON d.dt BETWEEN t.start_dt AND t.end_dt
WHERE  d.dt BETWEEN '2019-01-01' AND CURRENT_DATE()
       AND cid IN ( 0, 3, 15, 77, 71, 101, 186, 100 ); 
```
Push down distinct through `BroadcastNestedLoopJoin`:
```sql
SELECT DISTINCT t.at_prod_ref_id, d.week_beg_dt FROM
(  SELECT DISTINCT at_prod_ref_id, start_dt, end_dt
   FROM table1 WHERE cid IN ( 0, 3, 15, 77, 71, 101, 186, 100 )
) t
JOIN (
  SELECT DISTINCT dt, week_beg_dt FROM table2 WHERE dt BETWEEN '2019-01-01' AND current_date()
) d ON d.dt BETWEEN t.start_dt AND t.end_dt;
```

image

**Comment by wangyum** [comment]
@dongjoon-hyun @cloud-fan Impala support this feature: https://github.com/apache/impala/commit/827070b473c02da480f3a9d77c59f7031f9070c2

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1749]: Impala support this feature: apache/impala@827070b

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1746]: We cannot use column stats (`distinctCount`) to determine whether the optimizer push down it or not?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1746]: Do you mean only push down distinct when CBO is enabled and all column stats are exist and  there are duplicate values?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1746]: Yea, I meant so, `spark.sql.cbo.enabled` or `spark.sql.cbo.planStats.enabled`

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1746]: @cloud-fan What do you think? As far as I know, there is no column stats in our production. I don't know the status of other companies.

**Pull request closed without merging**